### PR TITLE
FIX-#2362: fix handling slices in 'DataFrame.__setitem__'

### LINKED ISSUE #2362: Error when using unhashable key in '__setitem__'
### Describe the problem

I get `TypeError: unhashable type: 'slice'` when trying to set value(s) to `Series` using slice.

### Source code / logs

```python
import modin.pandas as pd
import pandas

ps = pandas.Series([1,2,3])
ms = pd.Series([1,2,3])

ps[1:3] = 0
ps
0    1
1    0
2    0
dtype: int64

ms[1:3] = 0
TypeError: unhashable type: 'slice'
```

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2362 
- [x] tests added and passing

### CODE REVIEW

File: modin/pandas/base.py [hunk 1]
```diff
@@ -2708,7 +2708,11 @@ def __getitem__(self, key):
         else:
             return self._getitem(key)
 
-    def _getitem_slice(self, key):
+    def _setitem_slice(self, key: slice, value):
```
Comment (devin-petersohn) [inline modin/pandas/base.py:2711]: Please add a method level documentation to this method.

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
Looks good, I think it would be good to fully document the new method

File: modin/pandas/base.py (hunk 1 above)
Comment (dchigarev) [reply modin/pandas/base.py:2711]: added

File: modin/pandas/dataframe.py [hunk 2]
```diff
@@ -2038,24 +2041,23 @@ def __setitem__(self, key, value):
...
                     if key.shape != self.shape:
                         raise ValueError("Array must be same shape as DataFrame")
                     key = DataFrame(key, columns=self.columns)
                 return self.mask(key, value, inplace=True)
 
-            def setitem_without_string_columns(df):
+            def setitem_unhashable_key(df):
                 # Arrow makes memory-mapped objects immutable, so copy will allow them
                 # to be mutable again.
                 df = df.copy(True)
```
Comment (devin-petersohn) [inline modin/pandas/dataframe.py:2055]: Can we open a new issue to remove this copy? I'm not even sure these lines of code are tested.

**Review by devin-petersohn — APPROVED** [review]

**Pull request merged**